proptest/src/strategy/unions.rs

```diff
@@ -169,8 +173,121 @@ macro_rules! union_value_tree_body {
     }
 }
 
-impl<T : ValueTree> ValueTree for UnionValueTree<T> {
-    union_value_tree_body!(T::Value, access_vec);
+/// `ValueTree corresponding to `Union`.
+pub struct UnionValueTree<T : Strategy> {
+    options: Vec<LazyValueTree<T>>,
+    // This struct maintains the invariant that between function calls,
+    // `pick`, `min_pick` and `prev_pick` (if Some) always point to initialized
+    // trees.
+    pick: usize,
+    min_pick: usize,
+    prev_pick: Option<usize>,
+}
+
+macro_rules! lazy_union_value_tree_body {
+    ($typ:ty, $access:ident) => {
+        type Value = $typ;
+
+        fn current(&self) -> Self::Value {
+            $access!([] opt = self, self.pick, {
+                opt.as_inner().unwrap_or_else(||
+                    panic!(
+                        "value tree at self.pick = {} must be initialized",
+                        self.pick,
+                    )
+                ).current()
+            })
+        }
+
+        fn simplify(&mut self) -> bool {
+            let orig_pick = self.pick;
+            if $access!([mut] opt = self, orig_pick, {
+                opt.as_inner_mut().unwrap_or_else(||
+                    panic!(
+                        "value tree at self.pick = {} must be initialized",
+                        orig_pick,
+                    )
+                ).simplify()
+            }) {
+                self.prev_pick = None;
+                return true;
+            }
+
+            assert!(
+                self.pick >= self.min_pick,
+                "self.pick = {} should never go below self.min_pick = {}",
+                self.pick,
+                self.min_pick,
+            );
+            if self.pick == self.min_pick {
+                // No more simplification to be done.
+                return false;
+            }
+
+            // self.prev_pick is always a valid pick.
+            self.prev_pick = Some(self.pick);
+
+            while self.pick > self.min_pick {
+                self.pick -= 1;
+                let initialized = $access!([mut] opt = self, self.pick, {
+                    opt.maybe_init();
+                    opt.is_initialized()
+                });
+                if initialized {
+                    return true;
+                }
+            }
+
+            false
```

Comment:
Unless I'm misreading the code, if we get here, `self.pick` is an index to an uninitialised (failed) entry, which breaks `current()`.

Reply:
Yeah you're correct! I'll fix it with a temp variable and see if I can detect it with tests (think I need some tests with union variants that fail.)